File: apps/codemap-search/tests/fixtures/navigation/python/nested.py
from dataclasses import dataclass
from decimal import Decimal


@dataclass(frozen=True)
class OrderDto:
    total: Decimal


def traced(name: str):
    def decorate(fn):
        return fn

    return decorate


class CheckoutService:
    def __init__(self, repo, inventory, policy):
        self.repo = repo
        self.inventory = inventory
        self.policy = policy

    @traced("checkout.submit")
    def submit(self, lines: list[dict[str, Decimal]]) -> OrderDto:
        def line_total(line: dict[str, Decimal]) -> Decimal:
            return line["price"] * line["quantity"]

        total = sum(line_total(line) for line in lines)
        dto = OrderDto(total=total)

        self.policy.validate(dto)
        self.inventory.reserve(lines)
        self.repo.persist(dto)
        return dto

File: apps/codemap-search/tests/fixtures/navigation/typescript/nested.ts
import { OrderRepository } from "./repository";
import { createAudit } from "./audit";

type OrderLine = Readonly<{ price: string; quantity: number }>;
type OrderDto = Pick<OrderLine, "price" | "quantity"> & { total: number };

function Trace(): MethodDecorator {
  return () => undefined;
}

class OrderMapper {
  mapOrder(lines: readonly OrderLine[]): OrderDto {
    const total = lines.reduce((sum, line) => sum + Number(line.price) * line.quantity, 0);
    return { price: lines[0]?.price ?? "0", quantity: lines.length, total } as OrderDto;
  }
}

export class CheckoutService {
  constructor(
    private readonly repo: OrderRepository,
    private readonly mapper: OrderMapper,
  ) {}

  @Trace()
  async submit(lines: readonly OrderLine[]) {
    const dto = this.mapper.mapOrder(lines);
    const audit = createAudit(dto.total);

    await this.repo.persist(this.toDto(dto));
    audit.track("checkout-submitted");
  }

  private toDto(dto: OrderDto) {
    return { ...dto, total: dto.total satisfies number };
  }
}
